compiler-core/src/language_server/code_action.rs

```diff
@@ -7562,3 +7562,374 @@ impl<'ast> ast::visit::Visit<'ast> for RemovePrivateOpaque<'ast> {
+                    };
+                    pattern_with_variable(pattern_code)
+                })
+                .join(" | ");
+
+            let clause_code = self.code_at(clause.then.location());
+            let guard_code = match (outer_guard, &clause.guard) {
+                (Some(outer), Some(inner)) => {
+                    let mut outer_code = self.code_at(outer.location()).to_string();
+                    let mut inner_code = self.code_at(inner.location()).to_string();
+                    if ast::BinOp::And.precedence() > outer.precedence() {
+                        outer_code = format!("{{ {outer_code} }}")
+                    }
+                    if ast::BinOp::And.precedence() > inner.precedence() {
+                        inner_code = format!("{{ {inner_code} }}")
+                    }
+                    format!(" if {outer_code} && {inner_code}")
+                }
+                (None, Some(guard)) | (Some(guard), None) => {
+                    format!(" if {}", self.code_at(guard.location()))
+                }
+                (None, None) => "".into(),
+            };
+
+            new_clauses.push(format!("{new_patterns}{guard_code} -> {clause_code}"));
+        }
+
+        let pattern_nesting = self
+            .edits
+            .src_span_to_lsp_range(outer_clause_span)
+            .start
+            .character;
+        let indentation = " ".repeat(pattern_nesting as usize);
+
+        self.edits.replace(
+            outer_clause_span,
+            new_clauses.join(&format!("\n{indentation}")),
+        );
+
+        let mut action = Vec::with_capacity(1);
+        CodeActionBuilder::new("Collapse nested case")
+            .kind(CodeActionKind::REFACTOR_REWRITE)
+            .changes(self.params.text_document.uri.clone(), self.edits.edits)
+            .preferred(false)
+            .push_to(&mut action);
+        action
+    }
+
+    fn code_at(&self, span: SrcSpan) -> &str {
+        self.module
+            .code
+            .get(span.start as usize..span.end as usize)
+            .expect("location must be valid")
+    }
+}
+
+impl<'ast> ast::visit::Visit<'ast> for CollapseNestedCase<'ast> {
+    fn visit_typed_clause(&mut self, clause: &'ast ast::TypedClause) {
+        let ast::TypedClause {
+            pattern,
+            alternative_patterns,
+            then,
+            location,
+            guard,
+        } = clause;
+
+        // We can only flatten a case expression in a very specific case:
+        // - This pattern may be introducing multiple variables,
+        // - The expression following this branch must be a case, and
+        // - It must be matching on one of those variables
+        //
+        // For example:
+        //
+        // ```gleam
+        // Wibble(a, b, 1) -> case a { ... }
+        // Wibble(a, b, 1) -> case b { ... }
+        // ```
+
+        // We can't have any alternative patterns...
+        if alternative_patterns.is_empty()
```

Comment:
Rather than complex conditionals and progressively more indentation could you extract this all to a helper function and use early returns please 🙏